### Analyse data Titanic Disaster

### Try Model: Tensor Flow Neural Network

In [1]:
# setup absolute path to location of package Starts and config-file 
from inspect import getsourcefile
import os.path as path, sys
current_dir = path.dirname(path.abspath(getsourcefile(lambda:0)))
sys.path.insert(0, current_dir[:current_dir.rfind(path.sep)])

# import package Starts
from Starts.startmodskl import *
from Starts.startmodtf import *

local_kwargs {'data_path': './data/train.csv, ./data/titanic_no_nan.csv ,./data/test.csv', 'drop_obj_col': False, 'nan_drop_col': False, 'nan_drop_row': False, 'nan_zero': False, 'nan_mean': True, 'nan_mean_neighbors': False}


Using TensorFlow backend.


In [2]:
nonan_data = idata[1]
nonan_data.head()

Survived  Pclass  Sex        Age  Ticket      Fare  Title  FamilySize
0         0       3    1  29.699118     199    7.2292      5           0
1         1       3    0   0.750000     194   19.2583      4           3
2         0       3    1  29.699118     487    6.8583      5           0
3         1       1    1  36.000000      33  120.0000      5           3
4         0       2    1  29.699118     629   15.0500      5           0

In [3]:
len(nonan_data.columns)

8

In [4]:
len(np.unique(nonan_data['Survived'].values))

2

In [5]:
stf = StartModTFANN(n_classes=len(np.unique(nonan_data['Survived'].values)), dependent_label='Survived')
stf

In [6]:
# Setup neural network parameters
input_dim = len(nonan_data.columns)

new_param={'input_units': len(nonan_data.columns)-1, 'hidden_units':[2,1], 'output_units':1, 'optimizer':'Adam', 
           'activation_fn':'relu', 'learning_rate': 0.0001, 'steps':2000, 'batch_size':10, 
           'num_epochs': 5, 'feature_scl':True, 'loss_fn':'binary_crossentropy', 
           'drop_out':0.2, 'rec_drop_out':0.5, 'bias_initializer': 'random_uniform', 
           'depth_wise_initializer': 'random_uniform', 'seed': 10}

stf.update_parameters=new_param
stf.info_parameters()


Input_units: 7 

Hidden_units: [2, 1] 

Output_units: 1 

Optimizer: Adam 

Activation_function: relu 

Learning_Rate: 0.0001 

Training_Steps: 2000 

Batch_Size: 10 

Number_of_epochs: 5 

Feature_Scaling: True 

Loss_function: binary_crossentropy 

Drop_out: 0.2 

Recurrent_drop_out: 0.5 

Bias_Initializer: random_uniform 

Depth_wise_Initializer: random_uniform 

Seed: 10 



In [7]:
# Try method Keras Tensorflow to predict the Survivors (column 'Survived')
model, y_eval, y_pred = stf.keras_sequential(data=nonan_data)

(712, 7) (712,)
Epoch 1/5
712/712 [==============================] - 0s 607us/step - loss: 0.6909 - acc: 0.6138
Epoch 2/5
712/712 [==============================] - 0s 253us/step - loss: 0.6868 - acc: 0.6138
Epoch 3/5
712/712 [==============================] - 0s 258us/step - loss: 0.6834 - acc: 0.6138
Epoch 4/5
712/712 [==============================] - 0s 290us/step - loss: 0.6804 - acc: 0.6138
Epoch 5/5
712/712 [==============================] - 0s 115us/step

Model <keras.models.Sequential object at 0x7f3ce0e019e8>: Scores: 67.69%, Accuracy: 61.38%


In [8]:
# x_train.shape, x_eval.shape, y_train.shape, y_eval.shape

In [9]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])

# Compare the predicted value with true_value
len(y_eval), len(y_pred)

(179, 179)

In [10]:
confusion_matrix(y_eval, y_pred)

array([[112,   0],
       [ 67,   0]])

### Result: 
#### the correct predicted result 112
#### the wrong predicted result 67

In [11]:
# update parameters by adding more units in hidden layer and second hidden layer
new_param={'input_units': len(nonan_data.columns)-1, 'hidden_units':[10, 10, 10], 'output_units':1, 
           'optimizer':'Adam', 'activation_fn':'relu', 'learning_rate': 0.0001,
           'steps': 2000, 'batch_size': 10, 'num_epochs': 20, 'feature_scl': True, 
           'loss_fn':'binary_crossentropy', 'drop_out': 0.5, 'rec_drop_out': 0.6, 
           'bias_initializer': 'random_uniform', 'depth_wise_initializer': 'random_uniform', 'seed': 10}

stf.update_parameters=new_param
stf.info_parameters()


Input_units: 7 

Hidden_units: [10, 10, 10] 

Output_units: 1 

Optimizer: Adam 

Activation_function: relu 

Learning_Rate: 0.0001 

Training_Steps: 2000 

Batch_Size: 10 

Number_of_epochs: 20 

Feature_Scaling: True 

Loss_function: binary_crossentropy 

Drop_out: 0.5 

Recurrent_drop_out: 0.6 

Bias_Initializer: random_uniform 

Depth_wise_Initializer: random_uniform 

Seed: 10 



In [12]:
nonan_data.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'Ticket', 'Fare', 'Title',
       'FamilySize'],
      dtype='object')

In [13]:
nonan_data.head()

Survived  Pclass  Sex        Age  Ticket      Fare  Title  FamilySize
0         0       3    1  29.699118     199    7.2292      5           0
1         1       3    0   0.750000     194   19.2583      4           3
2         0       3    1  29.699118     487    6.8583      5           0
3         1       1    1  36.000000      33  120.0000      5           3
4         0       2    1  29.699118     629   15.0500      5           0

In [14]:
model, y_true, y_pred = stf.keras_sequential(data=nonan_data)

(712, 7) (712,)
Epoch 1/20
712/712 [==============================] - 1s 929us/step - loss: 0.6868 - acc: 0.5969
Epoch 2/20
712/712 [==============================] - 0s 396us/step - loss: 0.6494 - acc: 0.6138
Epoch 3/20
712/712 [==============================] - 0s 260us/step - loss: 0.6404 - acc: 0.6222
Epoch 4/20
712/712 [==============================] - 0s 263us/step - loss: 0.6280 - acc: 0.6713
Epoch 5/20
712/712 [==============================] - 0s 223us/step - loss: 0.6130 - acc: 0.6980
Epoch 6/20
712/712 [==============================] - 0s 322us/step - loss: 0.6161 - acc: 0.6966
Epoch 7/20
712/712 [==============================] - 0s 272us/step - loss: 0.6040 - acc: 0.6952
Epoch 8/20
712/712 [==============================] - 0s 247us/step - loss: 0.6043 - acc: 0.7065
Epoch 9/20
712/712 [==============================] - 0s 307us/step - loss: 0.6002 - acc: 0.7008
Epoch 10/20
712/712 [==============================] - 0s 292us/step - loss: 0.6001 - acc: 0.7079
Epoch 11/20
7

In [15]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])
confusion_matrix(y_true, y_pred)

array([[98, 14],
       [34, 33]])

### Result: 
#### the correct predicted result 131
#### the wrong predicted result 48
#### the accuracy has been improved from 61.38% to 72.61%

#### Notice: unnecessary feature (e.g. 'Ticket') might be removed before building model to improve the result overall